### PR TITLE
Windows support

### PR DESCRIPTION
**Pull request by JCapucho**
This PR allows `spotifyd` to run on windows as non-daemon and daemon-like, only the rodio-backend is supported and the config file is stored on:

```
%AppData%/Roaming/spotifyd/spotifyd.conf
```

All output is logged into a `.spotifyd.log` file on windows when running in both non-daemon and daemon mode although in daemon mode the startup is logged to the console before a detached subprocess with the same binary and arguments (with the exception that the argument `--no-daemon` is set) is started

### CODE REVIEW

**Comment by SirWindfield** [comment]
As this is (compared to other PRs) larger change I'll add it to the v1 milestone for now. This doesn't mean however that it won't get merged beforehand.

**Comment by MartB** [comment]
This seems to work perfectly fine on windows, just needs a rebase.

**Comment by JCapucho** [comment]
I rebased the branch and improved the logging now `--no-daemon` will use the terminal and the daemon will use the file but the file will be stored next to the config file instead of the current working directory this makes it easier to use windows services

**Comment by SirWindfield** [comment]
Thanks for that! I will merge this after I can update the librespot dependency.

**Comment by Plecra** [comment]
Just want to add that there's currently a runtime dependency on libogg (from `with-tremor`), and I wouldn't expect to find it on most windows systems. Users will have to do something about this if they want to use spotifyd.

**Comment by Plecra** [comment]
Also, why aren't we using the windows services API for this? It seems to fit the role best.

**Comment by JCapucho** [comment]
I'm longer working on this and don't plan on working on it since I finally got rid of windows on the last machine I used that had it, if my memory is right I didn't use services because they are hard to get right and require special permissions will this method doesn't

**Comment by ekx** [comment]
Tried this out today. Seems to work well so far... Found one Issue: When I switch the audio output devices while spotifyd is running it keeps playing on the old device. Tested on latest Windows 10

**Comment by CptPotato** [comment]
A small issue I ran into is that I can't get the `password_cmd` option to work on windows. Other than that it is working fine for me :+1:

**Comment by fightwithcrayon** [comment]
Seems that on Windows (10) the following commands work:

`cargo build --release --no-default-features --features rodio_backend`
`cargo run --no-default-features --features rodio_backend`

And this PR will build/run completely fine. However, it's not currently possible to install it using:

`cargo install --path . --locked`
`cargo install --path . --locked --no-default-features --features rodio_backend`

As it complains about:

`error: failed to run custom build command for 'alsa-sys v0.1.2'`

Which suggests to me it's ignoring the `rodio_backend` build flag.

**Comment by matteoturini** [comment]
Why is this dead? This is a great feature. I want to have spotify running HQ, without the desktop client.

**Comment by Brin-o** [comment]
Hi, just want to express my support for getting this pr in. It would be really cool to get a lightweight spotify client for windows.

**Comment by StaticPH** [comment]
I can confirm that with a bit of finagling, the code from the PR branch runs on Windows 8.1
To be clear, this means that the working build I have is based off the version 0.2.24 codebase, and _not_ a more recent version with JCapucho's changes patched in; I don't know if simply applying the changes to the current codebase would yield a working build, and I didn't particularly need anything from the 3.0.x release, so I never tried.

I found that I had to specify limited version ranges for `librespot`, `protobuf-codegen`, and `protobuf-codegen-pure` in Cargo.toml
```toml
[dependencies.librespot]
version = ">= 0.1.1, < 0.1.2"
default-features = false
features = ["with-tremor"]

[build-dependencies.protobuf-codegen]
version = "~2.8.1, < 2.9.0"

[build-dependencies.protobuf-codegen-pure]
version = "~2.8.1, < 2.9.0"
```
I _think_ I might have also needed to manually update the manifest and lockfiles for `librespot` accordingly, and build it with `--features="rodio-backend,with-tremor"`, but its been several weeks and its not clear from my command history if I did or not.

I _definitely_ had to regenerate/update the lockfile for `spotifyd` itself, making sure that it listed acceptable versions for `protobuf-codegen`, `protobuf-codegen-pure`, and the various `librespot` crates. I probably did that with `cargo generate-lockfile --offline`, but again, my command history doesn't indicate whether or not I did it by hand.

To build `spotifyd`, I had to use the MSVC toolchain rather than GNU toolchain, as I found that no matter what I did, the latter either failed to build, built an executable that simply didn't work, or built an executable I couldn't run; your results may differ.
The actual command I used is `rustup run stable-msvc cargo build --release --no-default-features --features="rodio_backend" --manifest ~/.cargo/git/checkouts/spotifyd-36f4565cca355e0b/d119fc3/Cargo.toml`
It may or may not also be possible with the right setup to build and run with the `dbus_keyring` and/or `dbus_mpris` features, but I didn't attempt that because just getting `spotifyd` to run at all was enough for me.

**Comment by vtermaat** [comment]
What needs to be done to get this merged?

**Comment by robinvd** [comment]
It should atleast be updated to the current master.

**Comment by lamaland** [comment]
Hi there, any help needed here? Would love to run this daemon on Windows.
Thanks for working on this.

**Comment by vtermaat** [comment]
I am using [librespot](https://github.com/librespot-org/librespot-java) instead

**Comment by CypherpunkSamurai** [comment]
Is this PR still active?

**Comment by DomingoMG** [comment]
Is something known?

**Comment by Katzenwerfer** [comment]
I believe it is currently preferred to use librespot (which was linked three comments above)

**Comment by brian6932** [comment]
I built this, and after playing with it for a bit, managed to get it to work.
For anyone on Windows that is getting:
```
Caught panic with message: attempted to zero-initialize type `librespot_tremor::tremor_sys::ov_callbacks`, which is invalid
```
Remove librespot's `with-tremor` feature flag in Cargo.toml
Specifically the 2nd thing here: https://github.com/Spotifyd/spotifyd/issues/719#issuecomment-774727927
The other methods of getting it to work did not work for me
Can someone please rebase this already, so that we don't have to do these annoying workarounds?

**Comment by slondr** [comment]
@Katzenwerfer @DomingoMG @CypherpunkSamurai 
This PR is fairly old, and has merge conflicts which aren't resolved, plus the author has already disclaimed it. Finally, I'm not sure any maintainer uses Windows to test this.

If someone could get this into a working, mergeable state, I may be able to throw together a Windows box to try it.

**Review by aadarshadhakalg — APPROVED** [review]

**Review by hackedXD — APPROVED** [review]
test

**Comment by CypherpunkSamurai** [comment]
We need two reviews for this PR to pass. I'll try cloning the branch and testing it out. If anyone else got time please do.

**Review by RakeshChowdhury — APPROVED** [review]
LGTM

**Comment by eladyn** [comment]
@RakeshChowdhury As has been pointed out in [this comment](https://github.com/Spotifyd/spotifyd/pull/602#issuecomment-1064702260), this PR currently is not mergeable and - up until now - no one has stepped up to fix the conflicts. So unless someone with interest shows up to get it working, this PR is unfortunately not going to happen. :/

**Comment by BKasin** [comment]
My branch here https://github.com/BKasin/spotifyd/tree/windows is rebased against the current commit of master, and seems to be working. Config file paths could definitely be handled better, and I'll have to research windows daemons/services if a better solution for that is desired. But as things stand it seems to work completely fine.

image

**Comment by BKasin** [comment]
@eladyn can probably close this now that #1219 is merged.

**Pull request closed without merging**
